### PR TITLE
Begin importing APIs under both Swift 3 and Swift 4 names, to produce errors

### PR DESCRIPTION
**Pull request by jrose-apple**
In order to ease the transition to Swift 4, we should import even renamed APIs with their Swift 3 name as well as their Swift 4 name, and use our existing rename machinery to correct the user when someone uses the wrong name (in either Swift 3 or Swift 4).

Very much a work in progress. Dependent on apple/swift-clang#53. rdar://problem/29170671

### CODE REVIEW

**Comment by jrose-apple** [comment]
@milseman, does this seem like the right approach? The last commit is probably the most interesting one for now.

File: lib/ClangImporter/ImportName.cpp [hunk 1]
```diff
@@ -576,12 +574,96 @@ determineCtorInitializerKind(const clang::ObjCMethodDecl *method) {
...
 }
 
+template <typename A>
+static bool matchesVersion(A *versionedAttr, ImportNameVersion version) {
+  clang::VersionTuple attrVersion = versionedAttr->getVersion();
+  if (attrVersion.empty()) {
+    // FIXME: This is brittle. The empty version represents "newest", as in
+    // "newer than the current version", but ImportNameVersion::Swift4 is
+    // a specific new version.
+    return version == ImportNameVersion::Swift4;
```
Comment (milseman) [inline lib/ClangImporter/ImportName.cpp]: I considered having a "Latest" version in ImportNameVersion. You could go with that, or at the very least a global in the ImportName header that is set to Swift4

File: lib/ClangImporter/ImportName.cpp [hunk 2]
```diff
@@ -576,12 +574,96 @@ determineCtorInitializerKind(const clang::ObjCMethodDecl *method) {
...
+      }
+
+      if (auto *removeAttr = dyn_cast<clang::SwiftVersionedRemovalAttr>(attr)) {
+        if (!matchesVersion(removeAttr, version))
+          continue;
+        if (removeAttr->getAttrKindToRemove() == clang::attr::SwiftName)
+          return nullptr;
+      }
+    }
+  }
```
Comment (milseman) [inline lib/ClangImporter/ImportName.cpp]: Would it simplify logic to put in a `return decl->getAttr<clang::SwiftNameAttr>()'` before this last brace, then you can drop the check later on?

File: lib/ClangImporter/ImportName.cpp (hunk 1 above)
Comment (milseman) [inline lib/ClangImporter/ImportName.cpp]: Actually, you do that further down. Any reason not to use that one?

Comment (jrose-apple) [inline lib/ClangImporter/ImportName.cpp]: Whoops, forgot I added that!

File: lib/ClangImporter/ImportName.cpp (hunk 2 above)
Comment (jrose-apple) [inline lib/ClangImporter/ImportName.cpp]: Ah, yeah, perhaps! Good catch that they're the same condition.

**Comment by jrose-apple** [comment]
Waiting until after Branch Day to land any of this.

**Comment by milseman** [comment]
Also, I approve of the brach name

**Comment by jrose-apple** [comment]
I'll put the follow-up work in later PRs; it's good to start getting this in.

@swift-ci Please test

**Comment by jrose-apple** [comment]
Rebased!

@swift-ci Please test

**Pull request merged**